mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala

```diff
@@ -512,6 +513,92 @@ object Vectors {
     squaredDistance
   }
 
+  private def dot(a : DenseVector, b : DenseVector) : Double = {
+    (a.toArray zip b.toArray).map(x => (x._1 * x._2)).sum
+  }
+
+  private def dot(a : SparseVector, b : DenseVector) : Double = {
+    (a.indices zip a.values).map(x => { b(x._1)*x._2 }).sum
```

Comment:
This probably won't work as it makes the sparse vector dense. In general I think these are slower than the Breeze impl. I also question the value of the Hadamard product as I've never actually had it come up.